torch/export/_trace.py

```diff
@@ -1674,7 +1674,24 @@ def override_getattribute_for_subclasses(args):
                     for k, (old_getattr, _) in tensor_type_to_old_getattribute.items():
                         k.__getattribute__ = old_getattr  # type: ignore[method-assign, attr-defined]
 
-            with ctx, override_getattribute_for_subclasses(flat_args):
+            @contextmanager
+            def _maybe_restore_grad_state():
+                """
+                When pre-dispatch export accidentally change grad state, we restore it back.
+                This can happen when we are calling torch._C._set_grad_enabled directly in the
+                forward.
+                """
+                old_state = torch.is_grad_enabled()
+                try:
+                    yield
+                finally:
+                    torch._C._set_grad_enabled(old_state)
+
+            with (
+                ctx,
+                override_getattribute_for_subclasses(flat_args),
+                _maybe_restore_grad_state(),
```

Comment:
Is it accurate that parameter mutation is supported in export (for all the flags/mode)? Might worth document all the semantic of mutations, e.g. input, buffer, parameters somewhere. Right now I don't have a clear mental model for them.